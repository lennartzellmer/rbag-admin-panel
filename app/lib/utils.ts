import { type ClassValue, clsx } from 'clsx'
import { twMerge } from 'tailwind-merge'
import { parseDate, today, getLocalTimeZone } from '@internationalized/date'
import type { EventSchema } from '~~/validation/eventSchema'

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs))
}

export function getEventStatusFromEvent(event: EventSchema) {
  if (!event.isPublished) return 'DRAFT'
  if (event.isCanceled) return 'CANCELED'
  if (!event.registration) return 'SAVE_THE_DATE'
  if (event.registration) {
    // Convert plain strings (or JS Date `.toISOString()` output) to CalendarDate
    const registrationStartDate = parseDate(event.registration.startDate.toISOString().substring(0, 10))
    const registrationEndDate = parseDate(event.registration.endDate.toISOString().substring(0, 10))

    // Get today's date in the local time zone as a CalendarDate
    const currentDate = today(getLocalTimeZone())

    // Compare currentDate to startDate and endDate
    if (currentDate.compare(registrationStartDate) < 0) {
      return 'REGISTRATION_SCHEDULED'
    }
    else if (currentDate.compare(registrationEndDate) < 0) {
      return 'REGISTRATION_OPEN'
    }
  }

  const eventStartDate = parseDate(event.startDate.toISOString().substring(0, 10))
  const eventEndDate = parseDate(event.endDate.toISOString().substring(0, 10))

  if (eventStartDate.compare(today(getLocalTimeZone())) > 0) {
    return 'UPCOMING'
  }
  if (eventEndDate.compare(today(getLocalTimeZone())) > 0) {
    return 'ONGOING'
  }
  return 'COMPLETED'
}

export type EventStatus = ReturnType<typeof getEventStatusFromEvent>
