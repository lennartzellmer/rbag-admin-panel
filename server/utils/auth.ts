import type { H3Event } from 'h3'
import { createError, getRequestHeader } from 'h3'
import { createRemoteJWKSet, decodeJwt, jwtVerify, type JWTPayload, type JWTVerifyGetKey } from 'jose'
import { ZITADEL_EMAIL_CLAIM, ZITADEL_ROLE_CLAIM, ZITADEL_USERID_CLAIM } from '~~/constants'

export type AuthUser = { id: string, email: string, roles: string[] }

type AuthConfig = {
  jwksUri: string
  issuer: string
  audience?: string | string[]
}

let remoteJwkSet: JWTVerifyGetKey | null = null

/**
 * Resolve and cache a process-wide RemoteJWKSet.
 * In a single-issuer setup, the JWKS URI does not change at runtime,
 * so we lazily initialize once and reuse for subsequent verifications.
 */
function resolveRemoteJwkSet(jwksUri: string): JWTVerifyGetKey {
  if (remoteJwkSet === null) {
    remoteJwkSet = createRemoteJWKSet(new URL(jwksUri))
  }

  return remoteJwkSet
}

/**
 * Resolve the email from the payload.
 */
function resolveEmail(payload: JWTPayload): string | null {
  if (typeof payload[ZITADEL_EMAIL_CLAIM] === 'string') {
    return payload[ZITADEL_EMAIL_CLAIM]
  }
  return null
}

/**
 * Resolve the user ID from the payload.
 */
function resolveUserId(payload: JWTPayload): string | null {
  if (typeof payload[ZITADEL_USERID_CLAIM] === 'string') {
    return payload[ZITADEL_USERID_CLAIM]
  }
  return null
}

/**
 * Extract the roles from the payload.
 */
function resolveRoles(payload: JWTPayload): string[] {
  const rawRoles = payload[ZITADEL_ROLE_CLAIM]

  if (!rawRoles || !Array.isArray(rawRoles)) {
    return []
  }

  const roles = new Set<string>()

  for (const role of rawRoles) {
    if (typeof role === 'string' && role.length > 0) {
      roles.add(role)
    }
  }

  return Array.from(roles).sort()
}

/**
 * Validate the Authorization bearer token from the incoming request.
 *
 * @param event H3Event carrying the HTTP request from which to read the Authorization header.
 * @param config Auth configuration including `jwksUri`, `issuer`, and optional `audience`.
 * @returns Promise that resolves to a discriminated union indicating validity and, on success, the authenticated user.
 */
export async function validateAuth(event: H3Event, config: AuthConfig): Promise<void> {
  const { jwksUri, issuer, audience } = config

  if (!jwksUri || !issuer || !audience) {
    console.error('Auth configuration is incomplete.')
    throw createError({
      statusCode: 401,
      statusMessage: 'Unauthorized',
      statusText: 'Auth configuration is incomplete.'
    })
  }

  // =============================================================================
  // Get the token from the Authorization header
  // =============================================================================

  const authorizationHeader = getRequestHeader(event, 'authorization')

  if (typeof authorizationHeader !== 'string' || !authorizationHeader.startsWith('Bearer ')) {
    throw createError({
      statusCode: 401,
      statusMessage: 'Unauthorized',
      statusText: 'Invalid Authorization header.'
    })
  }

  const token = authorizationHeader.slice('Bearer '.length).trim()

  if (token.length === 0) {
    throw createError({
      statusCode: 401,
      statusMessage: 'Unauthorized',
      statusText: 'Invalid Authorization header.'
    })
  }

  // =============================================================================
  // Verify the token
  // =============================================================================

  try {
    const key = resolveRemoteJwkSet(jwksUri)
    const verifyOptions = {
      issuer,
      audience: Array.isArray(audience) ? audience : [audience]
    }
    await jwtVerify(token, key, verifyOptions)
  }
  catch (error) {
    console.error('Failed to verify access token.', error)
    throw createError({
      statusCode: 401,
      statusMessage: 'Unauthorized',
      statusText: 'Failed to verify access token.'
    })
  }
}

/**
 * Decode the Authorization bearer token to extract user claims without verifying the signature.
 * This should only be used in trusted environments where the token has already been verified.
 *
 * @param event H3Event carrying the HTTP request from which to read the Authorization header.
 * @returns The authenticated user extracted from the token claims.
 * @throws An H3Error with status code 401 if the token is missing or invalid.
 */
export function extractAuthUser(event: H3Event): AuthUser {
  const authorizationHeader = getRequestHeader(event, 'authorization')

  if (typeof authorizationHeader !== 'string' || !authorizationHeader.startsWith('Bearer ')) {
    throw createError({
      statusCode: 400,
      statusMessage: 'Could not extract user from auth token',
      statusText: 'Invalid Authorization header.'
    })
  }

  const token = authorizationHeader.slice('Bearer '.length).trim()

  if (token.length === 0) {
    throw createError({
      statusCode: 400,
      statusMessage: 'Could not extract user from auth token',
      statusText: 'Invalid Authorization header.'
    })
  }

  try {
    const payload = decodeJwt(token)

    const userId = resolveUserId(payload)
    const email = resolveEmail(payload)
    const roles = resolveRoles(payload)

    if (!userId || !email || roles.length === 0) {
      throw createError({
        statusCode: 400,
        statusMessage: 'Could not extract user from auth token',
        statusText: 'Invalid Authorization header.'
      })
    }

    return {
      id: userId,
      email,
      roles
    }
  }
  catch (error) {
    console.error('Failed to decode access token.', error)
    throw createError({
      statusCode: 401,
      statusMessage: 'Could not extract user from auth token',
      statusText: 'Invalid Authorization header.'
    })
  }
}
