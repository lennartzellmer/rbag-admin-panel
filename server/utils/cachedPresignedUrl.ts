import type { Media } from '~~/shared/validation/mediaSchema'

export const cachedPresignedUrl = defineCachedFunction(async (media: Media) => {
  // =============================================================================
  // Generate presigned URL
  // =============================================================================

  const { minioClient } = useMinio()
  const runtimeConfig = useRuntimeConfig()
  const urlExpirationSeconds = runtimeConfig.storage.s3.uploadUrlExpirationSeconds
  const bucket = runtimeConfig.storage.s3.bucket

  const url = await minioClient.presignedGetObject(bucket, media.objectName, urlExpirationSeconds).catch((error) => {
    console.error('Failed to generate presigned URL:', error)
    throw createError({
      statusCode: 500,
      statusMessage: 'Failed to generate media URL'
    })
  })
  return {
    ...media,
    url
  }
}, {
  maxAge: (() => {
    const runtimeConfig = useRuntimeConfig()
    return Math.round(runtimeConfig.storage.s3.uploadUrlExpirationSeconds * 0.5)
  })(),
  name: 'minioPresignedUrl',
  getKey: (media: Media) => media.objectName
})
