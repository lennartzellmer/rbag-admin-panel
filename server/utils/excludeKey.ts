// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const excludeKey = <T extends object, U extends keyof any>(obj: T, key: U) => {
  const { [key]: _, ...newObj } = obj
  return newObj
}
