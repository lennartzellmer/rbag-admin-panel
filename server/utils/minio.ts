import { Client } from 'minio'

const useMinio = () => {
  const runtimeConfig = useRuntimeConfig()

  const minioClient = new Client({
    endPoint: runtimeConfig.storage.s3.endpoint,
    region: runtimeConfig.storage.s3.region,
    accessKey: runtimeConfig.storage.s3.accessKeyId,
    secretKey: runtimeConfig.storage.s3.secretAccessKey,
    useSSL: false
  })

  return minioClient
}
