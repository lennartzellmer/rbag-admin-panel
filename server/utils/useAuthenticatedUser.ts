import type { H3Event } from 'h3'
import { createError } from 'h3'
import { federatedUserSchema } from '../domain/user/validation'

export const useAuthenticatedUser = async (event: H3Event) => {
  const { user } = await getUserSession(event)

  const {
    success: isValidUser,
    error: userValidationError,
    data: validatedUserData
  } = await federatedUserSchema.safeParse(user)

  if (!isValidUser) {
    throw createError({
      statusCode: 400,
      statusMessage: 'Invalid user data',
      statusText: userValidationError?.message
    })
  }

  return validatedUserData
}
