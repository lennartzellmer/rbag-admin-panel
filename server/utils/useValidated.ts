import type z from 'zod'
import type { H3Event } from 'h3'
import { useSafeValidatedBody, useSafeValidatedParams, useSafeValidatedQuery } from 'h3-zod'

export async function useValidatedBody<T extends z.ZodTypeAny>(
  event: H3Event,
  schema: T
): Promise<z.infer<T>> {
  const {
    success: isValidBody,
    data: validatedBody,
    error: validationError
  } = await useSafeValidatedBody(
    event,
    schema
  )

  if (!isValidBody) {
    throw createError({
      statusCode: 400,
      statusMessage: 'Invalid event data',
      statusText: validationError?.message
    })
  }
  return validatedBody
}

export async function useValidatedQuery<T extends z.ZodTypeAny>(
  event: H3Event,
  schema: T
): Promise<z.infer<T>> {
  const {
    success: isValidQuery,
    data: validatedQuery,
    error: validationErrorQuery
  } = await useSafeValidatedQuery(event, schema)

  if (!isValidQuery) {
    throw createError({
      statusCode: 400,
      statusMessage: 'Invalid event data',
      statusText: validationErrorQuery?.message
    })
  }

  return validatedQuery
}

export async function useValidatedParams<T extends z.ZodTypeAny>(
  event: H3Event,
  schema: T
): Promise<z.infer<T>> {
  const {
    success: isValidParams,
    data: validatedParams,
    error: validationErrorParams
  } = await useSafeValidatedParams(event, schema)

  if (!isValidParams) {
    throw createError({
      statusCode: 400,
      statusMessage: 'Invalid event data',
      statusText: validationErrorParams?.message
    })
  }

  return validatedParams
}
