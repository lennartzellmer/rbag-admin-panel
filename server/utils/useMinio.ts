import { Client } from 'minio'

export const useMinio = () => {
  const runtimeConfig = useRuntimeConfig()

  const minioClient = new Client({
    endPoint: runtimeConfig.storage.s3.endpoint,
    port: runtimeConfig.storage.s3.port,
    region: runtimeConfig.storage.s3.region,
    accessKey: runtimeConfig.storage.s3.accessKeyId,
    secretKey: runtimeConfig.storage.s3.secretAccessKey,
    useSSL: false
  })

  async function ensureBucketExists(bucketName: string) {
    const exists = await minioClient.bucketExists(bucketName).catch(() => false)
    if (!exists) {
      await minioClient.makeBucket(bucketName, '')
    }
  }

  function getExtensionOf(name: string) {
    const idx = name.lastIndexOf('.')
    return idx >= 0 ? name.slice(idx + 1) : ''
  }

  return { minioClient,
    ensureBucketExists,
    getExtensionOf
  }
}
