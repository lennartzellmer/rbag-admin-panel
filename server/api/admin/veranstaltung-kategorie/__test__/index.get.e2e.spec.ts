import { beforeAll, describe, expect, test } from 'vitest'
import { $fetch } from '@nuxt/test-utils/e2e'
import type { PaginationResponseSchema } from '~~/validation/paginationQuerySchema'
import { setupCleanNuxtEnvironment } from '~~/test/utils/mongoMemoryServer'
import { createTestAuthHeader } from '~~/test/utils/jwt'

describe('VeranstaltungsKategorie GET - E2E Test', async () => {
  await setupCleanNuxtEnvironment()

  // Create a test JWT for the tests
  const authHeader = await createTestAuthHeader()

  beforeAll(async () => {
    await $fetch('/api/admin/veranstaltung-kategorie/create', {
      method: 'POST',
      body: {
        name: 'Kategorie A'
      },
      headers: { ...authHeader }
    })

    await $fetch('/api/admin/veranstaltung-kategorie/create', {
      method: 'POST',
      body: {
        name: 'Kategorie B'
      },
      headers: { ...authHeader }
    })
  })

  test('should return paginated veranstaltungsKategorien with metadata', async () => {
    const response = await $fetch<PaginationResponseSchema>('/api/admin/veranstaltung-kategorie', {
      query: {
        limit: '10',
        offset: '0'
      },
      headers: { ...authHeader }
    })

    expect(response).toMatchSnapshot({
      data: expect.arrayContaining([
        expect.objectContaining({
          name: expect.any(String)
        })
      ]),
      meta: {
        total: expect.any(Number),
        offset: 0,
        limit: 10
      }
    })

    expect(response).toBeDefined()
    expect(response.data).toEqual(expect.arrayContaining([
      expect.objectContaining({
        name: 'Kategorie A'
      }),
      expect.objectContaining({
        name: 'Kategorie B'
      })
    ]))

    expect(response.meta.total).toBeGreaterThanOrEqual(2)
  })

  test('should return 400 for invalid pagination query params', async () => {
    await expect(() => $fetch('/api/admin/veranstaltung-kategorie', {
      query: {
        limit: 'abc',
        offset: '0'
      },
      headers: { ...authHeader }
    })).rejects.toThrowError(expect.objectContaining({
      statusCode: 400
    }))
  })
})
