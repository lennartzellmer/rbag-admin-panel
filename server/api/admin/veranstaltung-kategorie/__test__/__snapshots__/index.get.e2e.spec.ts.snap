// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VeranstaltungsKategorie GET - E2E Test > should return paginated veranstaltungsKategorien with metadata 1`] = `
{
  "data": ArrayContaining [
    ObjectContaining {
      "name": Any<String>,
    },
  ],
  "meta": {
    "limit": 10,
    "offset": 0,
    "total": Any<Number>,
  },
}
`;
